src/NHibernate/Cache/StandardQueryCache.cs

```diff
@@ -510,23 +594,24 @@ public void Destroy()
 
 		private static ICacheAssembler[] GuessTypes(IList cacheable)
 		{
-			var colCount = (cacheable[0] as object[])?.Length ?? 1;
+			var colCount = (GetResultsEnumerable(cacheable).First() as object[])?.Length ?? 1;
 			var returnTypes = new ICacheAssembler[colCount];
 			if (colCount == 1)
 			{
-				foreach (var obj in cacheable)
+				foreach (var cached in GetResultsEnumerable(cacheable))
 				{
-					if (obj == null)
+					if (cached == null)
```

Comment:
Huh... We have different treatment for single column vs multi column result(object vs object[]).. So it's better to add test for colCount > 1 too...

Reply:
Added.